client/setec/store_test.go

```diff
@@ -0,0 +1,231 @@
+
+	t.Run("NewStore_missingURL", func(t *testing.T) {
+		st, err := setec.NewStore(ctx, setec.StoreConfig{})
+		if err == nil {
+			st.Close()
+			t.Errorf("Got %+v, want error", st)
+		} else {
+			t.Logf("Got expected error: %v", err)
+		}
+	})
+
+	t.Run("NewStore_emptySecrets", func(t *testing.T) {
+		st, err := setec.NewStore(ctx, setec.StoreConfig{Client: cli})
+		if err == nil {
+			st.Close()
+			t.Errorf("Got %+v, want error", st)
+		} else {
+			t.Logf("Got expected error: %v", err)
+		}
+	})
+
+	t.Run("NewStore_missingSecret", func(t *testing.T) {
+		ctx, cancel := context.WithTimeout(ctx, 50*time.Millisecond)
+		defer cancel()
+		st, err := setec.NewStore(ctx, setec.StoreConfig{
+			Client:  cli,
+			Secrets: []string{"nonesuch"},
+			Logf:    logger.Discard,
+		})
+		if err == nil {
+			st.Close()
+			t.Errorf("Got %+v, want error", st)
+		} else {
+			t.Logf("Got expected error: %v", err)
+		}
+	})
+
+	t.Run("NewStoreOK", func(t *testing.T) {
+		st, err := setec.NewStore(ctx, setec.StoreConfig{
+			Client:  cli,
+			Secrets: []string{"alpha", "bravo"},
+			Logf:    logger.Discard,
+		})
+		if err != nil {
+			t.Fatalf("Unexpected error: %v", err)
+		}
+		defer st.Close()
+
+		// We should be able to get a secret we requested.
+		if f := st.Secret("alpha"); f == nil {
+			t.Error("Lookup alpha: result is nil")
+		} else if got, want := string(f.Get()), "ok"; got != want {
+			t.Errorf("Lookup alpha: got %q, want %q", got, want)
+		}
+
+		// We should not be able to get a secret we didn't request.
+		if f := st.Secret("nonesuch"); f != nil {
+			t.Errorf("Lookup nonesuch: got %v, want nil", f)
+		}
+	})
+}
+
+func TestCachedStore(t *testing.T) {
+	const cacheData = `{"alpha":{"Value":"Zm9vYmFy","Version":100}}`
+
+	// Populate a memory cache with an "old" value of a secret.
+	mc := mustMemCache(t, cacheData)
+
+	// Connect to a service which has a newer value of the same secret, and
+	// verify that initially we see the cached value.
+	s := httptest.NewServer(testServer{secrets: map[string][]*api.SecretValue{
+		"alpha": {
+			{Value: []byte("bazquux"), Version: 200}, // a newer value
+			{Value: []byte("foobar"), Version: 100},  // the value in cache
+		}}})
+	defer s.Close()
+	cli := setec.Client{Server: s.URL, DoHTTP: s.Client().Do}
+
+	ctx := context.Background()
+	st, err := setec.NewStore(ctx, setec.StoreConfig{
+		Client:       cli,
+		Secrets:      []string{"alpha"},
+		Cache:        mc,
+		PollInterval: 50 * time.Millisecond,
+	})
+	if err != nil {
+		t.Fatalf("NewServer: unexpected error: %v", err)
+	}
+	defer st.Close()
+
+	alpha := st.Secret("alpha")
+	if alpha == nil {
+		t.Fatal("Lookup alpha: secret not found")
+	}
+
+	if got, want := string(alpha.Get()), "foobar"; got != want {
+		t.Fatalf("Lookup alpha: got %q, want %q", got, want)
+	}
+
+	// After the poller has had a chance to observe the new version, verify that
+	// we see it without having to update explicitly.
+	time.Sleep(100 * time.Millisecond)
```

Comment:
Yuck, my spidey sense is triggered. Does this mean the test is racy and dependent on the first part running within 45ms of the store being created? Our rubbish cloud CI is almost certain to violate that periodically.

Maybe you can use tailscale.com/tstest.Clock to plumb through virtual timers? That would let you control the triggering of the refresh explicitly, and make this reliable.

Reply:
I really quite dislike the fake clock package. It's just too much "abstraction".

It's not a read/write race, but it is indeed a timing race, though I chose these numbers so that even on a 0.1cpu throttled Azure runner it shouldn't fail often enough to matter.

Reply:
I added a hook to do this without too much violence. 🙂